ops/pebble.py

```diff
@@ -308,29 +317,97 @@ def __repr__(self):
+    """
+
+    def __init__(self, raw: Union[str, Dict] = None):
+        if isinstance(raw, str):
+            d = yaml.safe_load(raw) or {}
+        else:
+            d = raw or {}
+        self.summary = d.get('summary', '')
+        self.description = d.get('description', '')
+        self.services = {name: SetupService(name, service)
+                         for name, service in d.get('services', {}).items()}
+
+    def to_yaml(self) -> str:
+        """Convert this setup to its YAML representation."""
+        return yaml.dump(self.to_dict(), Dumper=_safe_dumper)
+
+    def to_dict(self) -> Dict:
+        """Convert this setup to a dict representation."""
+        fields = [
+            ('summary', self.summary),
+            ('description', self.description),
+            ('services', {name: service.to_dict() for name, service in self.services.items()})
+        ]
+        return {name: value for name, value in fields if value}
+
+    def __repr__(self) -> str:
+        return 'Setup({!r})'.format(self.to_dict())
+
+    __str__ = to_yaml
+
+
+class SetupService:
+    """Represents a service description in a Pebble setup layer."""
+
+    def __init__(self, name: str, raw: Dict = None):
+        self.name = name
+        raw = raw or {}
+        self.summary = raw.get('summary', '')
+        self.description = raw.get('description', '')
+        self.default = raw.get('default', '')
+        self.override = raw.get('override', '')
+        self.command = raw.get('command', '')
+        self.after = raw.get('after', [])
+        self.before = raw.get('before', [])
+        self.requires = raw.get('requires', [])
+        self.environment = raw.get('environment') or {}
+
+    def to_dict(self) -> Dict:
+        """Convert this service object to its dict representation."""
+        fields = [
+            ('summary', self.summary),
+            ('description', self.description),
+            ('default', self.default),
+            ('override', self.override),
+            ('command', self.command),
+            ('after', self.after),
+            ('before', self.before),
+            ('requires', self.requires),
+            ('environment', self.environment),
+        ]
+        return {name: value for name, value in fields if value}
+
+    def __repr__(self) -> str:
+        return 'SetupService({!r})'.format(self.to_dict())
+
+
+class PebbleClient:
     """Pebble API client."""
 
     def __init__(self, socket_path=None, opener=None, base_url='http://localhost', timeout=5.0):
-        """Initialize an instance. Defaults to Unix socket "$PEBBLE/.pebble.socket"."""
+        """Initialize a client instance.
+
+        Defaults to using a Unix socket at socket_path (which must be specified
+        unless a custom opener is provided).
+        """
         if opener is None:
+            if socket_path is None:
+                raise ValueError('no socket path provided')
             opener = self._get_default_opener(socket_path)
```

Comment:
Logic above doesn't seem to reflect the documentation. opener and socket_path both default to None, so it will error out instead of assuming default values.

Reply:
See above -- I had updated this. Now says:

> Defaults to using a Unix socket at socket_path (which must be specified
> unless a custom opener is provided).

Reply:
We should probably make it behave similar to the pebble client looking at the environment, but there's nothing pressing it to be done right now.

Reply:
Yep -- the testing Python CLI does that, but not at the `pebble.Client` level, at the CLI/main level.